Create  Unsent message RECdevelopment

areteruhiro opened the pull request:
### 確認項目

<!--
以下の項目を全て確認し、満たしている場合は
[ ] を [x] に書き換えてください。
該当部分以外は書き換えないでください。
-->

- [ ] <!-- Choice,multiple --> 動作確認済み
- [ ] <!-- Choice,multiple --> 誤字脱字無し

---
### 説明

取り消されたメッセージのリスト化

s1204IT reviewed (commented):
以下の事を確認してください

- [ ] パッケージID をオリジナル通りにする
- [ ] ファイル名を正しくする
- [ ] 多言語対応(英語等)する為に文字列をリソースから呼び出し

areteruhiro commented:
レビュー恐れ入ります。

現在、検討箇所があります。

現在の未読のまま閲覧　ボタンのところに、新しくボタンを作成して、後から変更出来るようにするかを悩んでいます。
現在このようなレイアウトで、ただでさえボタンが多すぎるような気もしています。
![Screenshot_20240831-233811.png](https://github.com/user-attachments/assets/3aa7f589-0639-48e2-a7fa-02901eeb5bf3)

また、現在の名前入力ボタンと、取り消しカウントボタンの２つが
重複？している？かどちらかしか作成されていないので、Chipさんに改善していただけるようなら、改善して頂きたいなと思っております。　

その他につきましては、今回の@s1204IT さんの
コメントにそって進めさせて頂きたいと思っております。よろしくお願いします。



　かなり話は変わりますが以下の非rootでのWear端末でのリンクについて、方法をread記載するかはお任せします。（QRコード認証廃止を加速させるおそれがあるため）
https://github.com/Chipppppppppp/LIME/issues/156#issuecomment-2322908803

areteruhiro commented:
あの、、本当に申し訳ないのですが、、これってどうやって消せますか、、
![image](https://github.com/user-attachments/assets/942c252d-6e2c-437d-bba3-9697350ccbd1)

areteruhiro commented:
ユーザー名どこかで取得できないかなと思ってたら、
他のテーブルに保存されてました。そりゃそうですよね
(contactsテーブル)
なので名前入力云々は必要ないです
コード書き換えますね

areteruhiro commented:
@s1204IT リソースからの呼び出しにすると、そのidみつからないよ！っておこられてしまうのですが、どうしたらいいですか？(もっと早く聞けばよかった)

areteruhiro commented:
`android.content.res.Resources$NotFoundException: String resource ID #0x7f100034
                                                                                                    	at android.content.res.Resources.getText(Resources.java:462)
                                                                                                    	at sjdeGERmjyxXPSjcQKfocSk.ources.getText(SourceFile:8)
                                                                                                    	at android.widget.TextView.setText(TextView.java:7578)
                                                                                                    	at io.test.hiro.lime.Main$1.afterHookedMethod(Main.java:76)
                                                                                                    	at tRAwuJss.SzE.Nvxrx.fDI.XposedBridge$LegacyApiSupport.handleAfter(Unknown Source:33)
                                                                                                    	at org.lsposed.lspd.impl.LSPosedBridge$NativeHooker.callback(Unknown Source:279)
                                                                                                    	at LSPHooker_.constructor(Unknown Source:14)
                                                                                                    	at java.lang.reflect.Constructor.newInstance0(Native Method)
                                                                                                    	at java.lang.reflect.Constructor.newInstance(Constructor.java:343)
                                                                                                    	at android.view.LayoutInflater.createView(LayoutInflater.java:742)
                                                                                                    	at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:894)
                                                                                                    	at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:849)
                                                                                                    	at android.view.LayoutInflater.rInflate(LayoutInflater.java:1011)
                                                                                                    	at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:972)
                                                                                                    	at android.view.LayoutInflater.inflate(LayoutInflater.java:570)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at jJnzE.JD.HcqaT.PkmmxqA.B.fEXj.HookBridge.invokeOriginalMethod(Native Method)
                                                                                                    	at org.lsposed.lspd.impl.LSPosedBridge$NativeHooker.callback(Unknown Source:187)
                                                                                                    	at LSPHooker_.inflate(Unknown Source:21)
                                                                                                    	at android.view.LayoutInflater.inflate(LayoutInflater.java:462)
                                                                                                    	at com.linecorp.line.chatlist.view.fragment.ChatListFragment$b.<init>(Unknown Source:22)
                                                                                                    	at com.linecorp.line.chatlist.view.fragment.ChatListFragment.onCreateView(Unknown Source:22)
                                                                                                    	at androidx.fragment.app.Fragment.performCreateView(Unknown Source:25)
                                                                                                    	at androidx.fragment.app.M.d(Unknown Source:174)
                                                                                                    	at androidx.fragment.app.M.i(Unknown Source:165)
                                                                                                    	at androidx.fragment.app.FragmentManager.D(Unknown Source:1225)
                                                                                                    	at androidx.fragment.app.FragmentManager.e0(Unknown Source:81)
                                                                                                    	at androidx.fragment.app.FragmentManager.C(Unknown Source:31)
                                                                                                    	at androidx.fragment.app.a.s(Unknown Source:10)
                                                                                                    	at androidx.fragment.app.J.d(Unknown Source:12)
                                                                                                    	at androidx.viewpager.widget.ViewPager.t(Unknown Source:667)
                                                                                                    	at androidx.viewpager.widget.ViewPager.s(Unknown Source:2)
                                                                                                    	at androidx.viewpager.widget.ViewPager.onMeasure(Unknown Source:186)
                                                                                                    	at android.view.View.measure(View.java:27866)
                                                                                                    	at android.widget.RelativeLayout.measureChildHorizontal(RelativeLayout.java:735)
                                                                                                    	at android.widget.RelativeLayout.onMeasure(RelativeLayout.java:481)
                                                                                                    	at android.view.View.measure(View.java:27866)
                                                                                                    	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:7028)
                                                                                                    	at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1608)
                                                                                                    	at android.widget.LinearLayout.measureHorizontal(LinearLayout.java:1240)
                                                                                                    	at android.widget.LinearLayout.onMeasure(LinearLayout.java:723)
                                                                                                    	at android.view.View.measure(View.java:27866)
                                                                                                    	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:7028)
                                                                                                    	at android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
                                                                                                    	at androidx.appcompat.widget.ContentFrameLayout.onMeasure(Unknown Source:159)
                                                                                                    	at android.view.View.measure(View.java:27866)
                                                                                                    	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:7028)
                                                                                                    	at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1608)
                                                                                                    	at android.widget.LinearLayout.measureVertical(LinearLayout.java:878)
                                                                                                    	at android.widget.LinearLayout.onMeasure(LinearLayout.java:721)
                                                                                                    	at android.view.View.measure(View.java:27866)
                                                                                                    	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:7028)
                                                                                                    	at android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
                                                                                                    	at android.view.View.measure(View.java:27866)
                                                                                                    	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:7028)
                                                                                                    	at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1608)
                                                                                                    	at android.widget.LinearLayout.measureVertical(LinearLayout.java:878)
                                                                                                    	at android.widget.LinearLayout.onMeasure(LinearLayout.java:721)
                                                                                                    	at android.view.View.measure(View.java:27866)
                                                                                                    	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:7028)
2024-09-09 20:05:10.068 29328-29328 LSPosed-Bridge          jp.naver.line.android                E  	at android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
                                                                                                    	at com.android.internal.policy.DecorView.onMeasure(DecorView.java:768)
                                                                                                    	at android.view.View.measure(View.java:27866)
                                                                                                    	at android.view.ViewRootImpl.performMeasure(ViewRootImpl.java:4486)
                                                                                                    	at android.view.ViewRootImpl.measureHierarchy(ViewRootImpl.java:3012)
                                                                                                    	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:3333)
                                                                                                    	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:2718)
                                                                                                    	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:9937)
                                                                                                    	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1406)
                                                                                                    	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1415)
                                                                                                    	at android.view.Choreographer.doCallbacks(Choreographer.java:1015)
                                                                                                    	at android.view.Choreographer.doFrame(Choreographer.java:945)
                                                                                                    	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1389)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:959)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:100)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:232)
                                                                                                    	at android.os.Looper.loop(Looper.java:317)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:8592)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:580)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:878)`

s1204IT commented:
`Resources$NotFoundException`ですよね。  
それ僕も別のプロジェクトでなって困ってるんですよね。

僕も検証中なので、進捗があれば報告します。

areteruhiro commented:
恐らくなんですが、フックしようとしているアプリ(今回の場合line)のリソースIDから取得してしまってResources$NotFoundExceptionが発生したのだと思います。
なので取得する先を指定してあげることで解決できました

areteruhiro commented:
他のメゾットで、同じようにしようとするとエラー出るのでまだ、直さないとだめですね、、

areteruhiro commented:
横スクロールの対応がまだです。

'package io.test.hiro.lime;

import java.io.IOException;
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.File;

import android.app.AlertDialog;
import android.app.Application;
import android.content.Context;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.graphics.Color;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.HorizontalScrollView;
import android.widget.ListView;
import android.widget.RelativeLayout;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;

import java.io.FileReader;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStreamReader;


import de.robv.android.xposed.IXposedHookLoadPackage;
import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XposedBridge;
import de.robv.android.xposed.XposedHelpers;
import de.robv.android.xposed.callbacks.XC_LoadPackage;
import io.test.hiro.lime.R;


public class Main implements IXposedHookLoadPackage {
    public static final String Main_file = "UNSENT_REC.txt";
    public static final String Main_backup = "BackUpFile.txt";

    private SQLiteDatabase db;
    private static final int MAX_RETRIES = 3;
    private Context context; // メンバ変数として context を追加

    SQLiteDatabase db1 = null;
    SQLiteDatabase db2 = null;


    @Override
    public void handleLoadPackage(XC_LoadPackage.LoadPackageParam lparam) throws Throwable {

        if (!lparam.packageName.equals("jp.naver.line.android"))
            return;
        XposedBridge.hookAllConstructors(
                lparam.classLoader.loadClass("jp.naver.line.android.common.view.listview.PopupListView"),
                new XC_MethodHook() {
                    @Override
                    protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                        ViewGroup viewGroup = (ViewGroup) param.thisObject;
                        Context context = viewGroup.getContext();
                        RelativeLayout container = new RelativeLayout(context);
                        RelativeLayout.LayoutParams containerParams = new RelativeLayout.LayoutParams(
                                RelativeLayout.LayoutParams.MATCH_PARENT, RelativeLayout.LayoutParams.WRAP_CONTENT);
                        container.setLayoutParams(containerParams);
                        Button openFileButton = new Button(context);
                        openFileButton.setText("確認済みのメッセージ");
                        openFileButton.setTextSize(12);
                        openFileButton.setTextColor(Color.BLACK);
                        openFileButton.setId(View.generateViewId());
                        RelativeLayout.LayoutParams buttonParams = new RelativeLayout.LayoutParams(
                                RelativeLayout.LayoutParams.WRAP_CONTENT, RelativeLayout.LayoutParams.WRAP_CONTENT);
                        buttonParams.addRule(RelativeLayout.CENTER_HORIZONTAL);
                        container.addView(openFileButton, buttonParams);
                        Button clearFileButton = new Button(context);
                        clearFileButton.setText("メッセージを削除");
                        clearFileButton.setTextSize(12);
                        clearFileButton.setTextColor(Color.RED);
                        clearFileButton.setId(View.generateViewId());
                        RelativeLayout.LayoutParams clearButtonParams = new RelativeLayout.LayoutParams(
                                RelativeLayout.LayoutParams.WRAP_CONTENT, RelativeLayout.LayoutParams.WRAP_CONTENT);
                        clearButtonParams.addRule(RelativeLayout.BELOW, openFileButton.getId());
                        clearButtonParams.addRule(RelativeLayout.CENTER_HORIZONTAL);
                        container.addView(clearFileButton, clearButtonParams);


                        openFileButton.setOnClickListener(v -> {
                            File backupFile = new File(context.getFilesDir(), Main_backup);
                            if (!backupFile.exists()) {
                                try {
                                    backupFile.createNewFile();
                                } catch (IOException e) {
                                    e.printStackTrace();
                                    Toast.makeText(context,"ファイルを作成できませんでした", Toast.LENGTH_SHORT).show();
                                    return;
                                }
                            }
                            if (backupFile.length() > 0) {

                                try {
                                    StringBuilder output = new StringBuilder();
                                    BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream(backupFile)));
                                    String line;
                                    while ((line = reader.readLine()) != null) {
                                        output.append(line).append("\n");
                                    }
                                    reader.close();

                                    // カスタムViewを作成
                                    HorizontalScrollView horizontalScrollView = new HorizontalScrollView(context); // 横スクロール用のScrollView
                                    ScrollView verticalScrollView = new ScrollView(context); // 縦スクロール用のScrollView

                                    TextView textView = new TextView(context);
                                    textView.setText(output.toString());

                                    // TextViewの設定（改行を防ぎ、横スクロールをサポート）
                                    textView.setMaxLines(Integer.MAX_VALUE); // 最大行数を設定（実質無制限）
                                    textView.setHorizontallyScrolling(true); // 横スクロールを有効にする
                                    textView.setHorizontalScrollBarEnabled(true); // 横スクロールバーを表示する

                                    // スクロールビューにTextViewを追加
                                    horizontalScrollView.addView(textView); // 横スクロールをサポートするScrollViewにTextViewを追加
                                    verticalScrollView.addView(horizontalScrollView); // 縦スクロールをサポートするScrollViewに横ScrollViewを追加

                                    // AlertDialogを作成
                                    AlertDialog.Builder builder = new AlertDialog.Builder(context);
                                    builder.setTitle("バックアップ")
                                            .setView(verticalScrollView)  // カスタムビューをセット
                                            .setPositiveButton("OK", null)
                                            .create()
                                            .show();
                                } catch (IOException e) {
                                    e.printStackTrace();
                                    Toast.makeText(context, "BackUpFile.txtの読み取りに失敗しました", Toast.LENGTH_SHORT).show();
                                }

                            } else {
                                Toast.makeText(context, "何もバックアップされていません", Toast.LENGTH_SHORT).show();
                            }
                        });
                        clearFileButton.setOnClickListener(v -> {
                            new AlertDialog.Builder(context)
                                    .setTitle("確認")
                                    .setMessage("メッセージを削除しますか")
                                    .setPositiveButton("yes", (dialog, which) -> {
                                        File backupFile = new File(context.getFilesDir(), Main_backup);
                                        if (backupFile.exists()) {
                                            try {
                                                BufferedWriter writer = new BufferedWriter(new FileWriter(backupFile));
                                                writer.write("");
                                                writer.close();
                                                Toast.makeText(context, "ファイルの内容が削除されました", Toast.LENGTH_SHORT).show();
                                            } catch (IOException e) {
                                                e.printStackTrace();
                                                Toast.makeText(context, "ファイルの削除に失敗しました", Toast.LENGTH_SHORT).show();
                                            }
                                        } else {
                                            Toast.makeText(context,"ファイルが見つかりません", Toast.LENGTH_SHORT).show();
                                        }
                                    })
                                    .setNegativeButton("no", null)
                                    .create()
                                    .show();
                        });


                        ((ListView) viewGroup.getChildAt(0)).addFooterView(container);
                    }
                }
        );


        XposedHelpers.findAndHookMethod(
                "com.linecorp.line.chatlist.view.fragment.ChatListPageFragment",
                lparam.classLoader, "onCreateView",
                LayoutInflater.class, ViewGroup.class, android.os.Bundle.class,
                new XC_MethodHook() {
                    @Override
                    protected void afterHookedMethod(MethodHookParam param) throws Throwable {

                        View rootView = (View) param.getResult();
                        Context context = rootView.getContext();


                        File originalFile = new File(context.getFilesDir(), Main_file);
                        if (!originalFile.exists()) {
                            try {
                                originalFile.createNewFile();
                            } catch (IOException e) {
                                e.printStackTrace();
                                Toast.makeText(context,
                                        "ファイルを作成できませんでした", Toast.LENGTH_SHORT).show();
                                return;
                            }
                        }
                        if (originalFile.length() > 0) {
                            int lineCount = countLinesInFile(originalFile);

                            if (lineCount > 0) {
                                Button button = new Button(context);
                                button.setText(Integer.toString(lineCount));
                                int buttonId = View.generateViewId();
                                button.setId(buttonId);  // IDを設定
                                RelativeLayout.LayoutParams buttonParams = new RelativeLayout.LayoutParams(
                                        RelativeLayout.LayoutParams.WRAP_CONTENT, RelativeLayout.LayoutParams.WRAP_CONTENT);
                                buttonParams.addRule(RelativeLayout.ALIGN_PARENT_BOTTOM);
                                button.setLayoutParams(buttonParams);

                                // SharedPreferencesでフラグの状態を取得
                                SharedPreferences prefs = context.getSharedPreferences("app_prefs", Context.MODE_PRIVATE);
                                boolean messageShown = prefs.getBoolean("messageShown", false); // デフォルトはfalse

                                button.setOnClickListener(v -> {
                                    try {
                                        StringBuilder output = new StringBuilder();
                                        StringBuilder updatedContent = new StringBuilder();
                                        BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream(originalFile)));
                                        String line;
                                        boolean showToast = false;  // トーストを表示するかどうか

                                        while ((line = reader.readLine()) != null) {
                                            if (line.contains("No content") || line.contains("No name")) {
                                                // メッセージが表示されていない場合にだけ表示
                                                if (!messageShown) {
                                                    showToast = true;
                                                    // フラグをSharedPreferencesに保存
                                                    SharedPreferences.Editor editor = prefs.edit();
                                                    editor.putBoolean("messageShown", true);
                                                    editor.apply();
                                                }
                                                // 取得できなかったメッセージは表示せず、削除のため更新しない
                                                continue;
                                            }
                                            output.append(line).append("\n");
                                            updatedContent.append(line).append("\n"); // 有効な行は保持
                                        }
                                        reader.close();

                                        // トーストメッセージの表示
                                        if (showToast) {
                                            Toast.makeText(context, "正しく取得できませんでした。アプリを再起動してください", Toast.LENGTH_SHORT).show();
                                        }



                                        // カスタムViewを作成
                                        HorizontalScrollView horizontalScrollView = new HorizontalScrollView(context); // 横スクロール用のScrollView
                                        ScrollView verticalScrollView = new ScrollView(context); // 縦スクロール用のScrollView

                                        TextView textView = new TextView(context);
                                        textView.setText(output.toString());

// TextViewの設定（改行を防ぎ、横スクロールをサポート）
                                        textView.setMaxLines(Integer.MAX_VALUE); // 最大行数を設定（実質無制限）
                                        textView.setHorizontallyScrolling(true); // 横スクロールを有効にする
                                        textView.setHorizontalScrollBarEnabled(true); // 横スクロールバーを表示する

// スクロールビューにTextViewを追加
                                        horizontalScrollView.addView(textView); // 横スクロールをサポートするScrollViewにTextViewを追加
                                        verticalScrollView.addView(horizontalScrollView); // 縦スクロールをサポートするScrollViewに横ScrollViewを追加

// AlertDialogを作成
                                        AlertDialog.Builder builder = new AlertDialog.Builder(context);
                                        builder.setTitle("削除されたメッセージ")
                                                .setView(verticalScrollView)  // カスタムビューをセット
                                                .setPositiveButton("ok", (dialog, which) -> {
                                                    try {
                                                        File backupFile = new File(context.getFilesDir(), Main_backup);
                                                        BufferedWriter writer = new BufferedWriter(new FileWriter(backupFile, true));
                                                        writer.write(output.toString());
                                                        writer.close();

                                                        // originalFileの内容をクリアしてバックアップ
                                                        BufferedWriter clearWriter = new BufferedWriter(new FileWriter(originalFile));
                                                        clearWriter.close();

                                                        Toast.makeText(context,
                                                                "内容がバックアップファイルに移動されました", Toast.LENGTH_SHORT).show();

                                                        // フラグをリセット
                                                        SharedPreferences.Editor editor = prefs.edit();
                                                        editor.putBoolean("messageShown", false);
                                                        editor.apply();

                                                    } catch (IOException e) {
                                                        e.printStackTrace();
                                                        Toast.makeText(context, "ファイルの移動に失敗しました", Toast.LENGTH_SHORT).show();
                                                    }

                                                    // ボタンの親からボタンを削除
                                                    if (button.getParent() instanceof ViewGroup) {
                                                        ViewGroup parent = (ViewGroup) button.getParent();
                                                        parent.removeView(button);
                                                    }
                                                })
                                                .create()
                                                .show();

                                    } catch (IOException e) {
                                        e.printStackTrace();
                                        Toast.makeText(context, "BackUpFile.txtの読み取りに失敗しました", Toast.LENGTH_SHORT).show();
                                    }
                                });

                                if (rootView instanceof ViewGroup) {
                                    ViewGroup viewGroup = (ViewGroup) rootView;
                                    viewGroup.addView(button);

                                    View existingButton = viewGroup.findViewById(buttonId);
                                    if (existingButton != null) {

                                    }
                                }
                            }
                        }








                    }
                }
        );


        XposedBridge.hookAllMethods(Application.class, "onCreate", new XC_MethodHook() {
            @Override
            protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                Application appContext = (Application) param.thisObject;

                File dbFile1 = appContext.getDatabasePath("naver_line");

                File dbFile2 = appContext.getDatabasePath("contact");


                if (dbFile1.exists() && dbFile2.exists()) {

                    SQLiteDatabase.OpenParams.Builder builder1 = new SQLiteDatabase.OpenParams.Builder();
                    builder1.addOpenFlags(SQLiteDatabase.OPEN_READWRITE);
                    SQLiteDatabase.OpenParams dbParams1 = builder1.build();

                    SQLiteDatabase.OpenParams.Builder builder2 = new SQLiteDatabase.OpenParams.Builder();
                    builder2.addOpenFlags(SQLiteDatabase.OPEN_READWRITE);
                    SQLiteDatabase.OpenParams dbParams2 = builder2.build();

                    db1 = SQLiteDatabase.openDatabase(dbFile1, dbParams1);
                    db2 = SQLiteDatabase.openDatabase(dbFile2, dbParams2);

                    hookMessageDeletion(lparam, appContext, db1, db2);
                    resolveUnresolvedIds(lparam, appContext, db1, db2);

                }

            }
        });






    }

    private String queryDatabase(SQLiteDatabase db, String query, String... selectionArgs) {
        if (db == null) {
            XposedBridge.log("Database is not initialized.");
            return null;
        }
        Cursor cursor = db.rawQuery(query, selectionArgs);
        String result = null;
        if (cursor.moveToFirst()) {
            result = cursor.getString(0);
        }
        cursor.close();
        return result;
    }

    private int countLinesInFile(File file) {
        int count = 0;
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream(file)))) {
            String line;
            while ((line = reader.readLine()) != null) {
                if (!(line.contains("No content") || line.contains("No name"))) {
                    count++; // 条件に合致しない行だけをカウント
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return count;
    }


    private void hookMessageDeletion(XC_LoadPackage.LoadPackageParam lparam, Context context, SQLiteDatabase db1, SQLiteDatabase db2) {
        try {
            Class<?> hookTarget = lparam.classLoader.loadClass("org.apache.thrift.k");
            XposedBridge.hookAllMethods(hookTarget, "a", new XC_MethodHook() {


                @Override
                protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                    String paramValue = param.args[1].toString();
                    if (paramValue.contains("type:NOTIFIED_DESTROY_MESSAGE,")) {
                        processMessage(paramValue, context, db1, db2);
                    }
                }
            });
        } catch (ClassNotFoundException e) {
            XposedBridge.log("Class not found: " + e.getMessage());
        }
    }

    private void processMessage(String paramValue, Context context, SQLiteDatabase db1, SQLiteDatabase db2) {
        String unresolvedFilePath = context.getFilesDir() + "/UnresolvedIds.txt";

        String[] operations = paramValue.split("Operation\\(");
        for (String operation : operations) {
            if (operation.trim().isEmpty()) continue;
            String revision = null;
            String createdTime = null;
            String type = null;
            String from = null;
            String to = null;
            String param12 = null;
            String param22 = null;
            String operationContent = null;
            String serverId = null;
            String talkId = null;

            String[] parts = operation.split(",");
            for (String part : parts) {
                part = part.trim();
                if (part.startsWith("param1:")) {
                    talkId = part.substring("param1:".length()).trim();
                } else if (part.startsWith("param2:")) {
                    serverId = part.substring("param2:".length()).trim();
                } else if (part.startsWith("revision:")) {
                    revision = part.substring("revision:".length()).trim();
                } else if (part.startsWith("createdTime:")) {
                    createdTime = part.substring("createdTime:".length()).trim();
                } else if (part.startsWith("type:")) {
                    type = part.substring("type:".length()).trim();
                } else if (part.startsWith("from:")) {
                    from = part.substring("from:".length()).trim();
                } else if (part.startsWith("to:")) {
                    to = part.substring("to:".length()).trim();
                } else if (part.startsWith("contentMetadata:")) {
                    param12 = part.substring("contentMetadata:".length()).trim();
                } else if (part.startsWith("operationContent:")) {
                    operationContent = part.substring("operationContent:".length()).trim();
                }
            }

            if (serverId != null && talkId != null) {
                String content = queryDatabase(db1, "SELECT content FROM chat_history WHERE server_id=?", serverId);

                String imageCheck = queryDatabase(db1, "SELECT attachement_image FROM chat_history WHERE server_id=?", serverId);
                String timeEpochStr = queryDatabase(db1, "SELECT created_time FROM chat_history WHERE server_id=?", serverId);
                String timeFormatted = formatMessageTime(timeEpochStr);

                String groupName = queryDatabase(db1, "SELECT name FROM groups WHERE id=?", talkId);
                String media = queryDatabase(db1, "SELECT attachement_type FROM chat_history WHERE server_id=?", serverId);

                // contactデータベースからtalkNameを取得
                String talkName = queryDatabase(db2, "SELECT profile_name FROM contacts WHERE mid=?", talkId);

                String name = (groupName != null ? groupName + ":" + talkName : (talkName != null ? talkName : "No Name" + ":" + ":" + "talkId" + talkId));

                if (timeEpochStr == null) {
                    saveUnresolvedIds(serverId, talkId, unresolvedFilePath);
                }

                // groupNameが取得できた場合、from_midとsender_nameを取得する
                String from_mid = null;
                String sender_name = null;
                if (groupName != null) {
                    from_mid = queryDatabase(db1, "SELECT from_mid FROM chat_history WHERE server_id=?", serverId);
                    if (from_mid != null) {
                        sender_name = queryDatabase(db2, "SELECT profile_name FROM contacts WHERE mid=?", from_mid);
                    }
                }

                // sender_nameが取得できた場合、nameを更新する
                if (sender_name != null) {
                    name = groupName + ": " + sender_name;
                }

                String mediaDescription = "";
                if (media != null) {
                    switch (media) {
                        case "7":
                            mediaDescription = "スタンプ";
                            break;
                        case "1":
                            mediaDescription = "写真";
                            break;
                        case "2":
                            mediaDescription = "動画";
                            break;
                        default:
                            mediaDescription = "";
                            break;
                    }
                }

                String logEntry = (timeFormatted != null ? timeFormatted : "No Time: ")
                        + name
                        + ": "
                        + ((content != null) ? content : (mediaDescription.isEmpty() ? "No content:" + serverId : ""))
                        + mediaDescription;

                File fileToWrite = new File(context.getFilesDir(), Main_file);

                try (BufferedWriter writer = new BufferedWriter(new FileWriter(fileToWrite, true))) {
                    writer.write(logEntry);
                    writer.newLine();
                } catch (IOException e) {
                    XposedBridge.log("IOException occurred while writing to file: " + e.getMessage());
                }
            }
        }
    }



    private void saveUnresolvedIds(String serverId, String talkId, String unresolvedFilePath) {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(unresolvedFilePath, true))) {
            writer.write("serverId:" + serverId + ",talkId:" + talkId);
            writer.newLine();
        } catch (IOException e) {
            XposedBridge.log("IOException occurred while saving unresolved IDs: " + e.getMessage());
        }
    }

    private void resolveUnresolvedIds(XC_LoadPackage.LoadPackageParam lparam, Context context, SQLiteDatabase db1, SQLiteDatabase db2) {
        String unresolvedFilePath = context.getFilesDir() + "/UnresolvedIds.txt";

        File unresolvedFile = new File(unresolvedFilePath);
        File testFile = new File(context.getFilesDir(), Main_file);

        if (!unresolvedFile.exists()) return;

        try (BufferedReader reader = new BufferedReader(new FileReader(unresolvedFile));
             BufferedWriter testWriter = new BufferedWriter(new FileWriter(testFile, true))) {

            String line;
            while ((line = reader.readLine()) != null) {
                String[] parts = line.split(",");
                String serverId = parts[0].split(":")[1];
                String talkId = parts[1].split(":")[1];

                String content = queryDatabase(db1, "SELECT content FROM chat_history WHERE server_id=?", serverId);
                String imageCheck = queryDatabase(db1, "SELECT attachement_image FROM chat_history WHERE server_id=?", serverId);
                String timeEpochStr = queryDatabase(db1, "SELECT created_time FROM chat_history WHERE server_id=?", serverId);
                String timeFormatted = formatMessageTime(timeEpochStr);
                String groupName = queryDatabase(db1, "SELECT name FROM groups WHERE id=?", talkId);
                String media = queryDatabase(db1, "SELECT attachement_type FROM chat_history WHERE server_id=?", serverId);
                // contactデータベースからtalkNameを取得
                String talkName = queryDatabase(db2, "SELECT profile_name FROM contacts WHERE mid=?", talkId);

                String name = (groupName != null ? groupName : (talkName != null ? talkName : "No Name" + ":" + ":" + "talkId" + talkId));


                // groupNameが取得できた場合、from_midとsender_nameを取得する
                String from_mid = null;
                String sender_name = null;
                if (groupName != null) {
                    from_mid = queryDatabase(db1, "SELECT from_mid FROM chat_history WHERE server_id=?", serverId);
                    if (from_mid != null) {
                        sender_name = queryDatabase(db2, "SELECT profile_name FROM contacts WHERE mid=?", from_mid);
                    }
                }

                // sender_nameが取得できた場合、nameを更新する
                if (sender_name != null) {
                    name = groupName + ": " + sender_name;
                }


                String logEntry = (timeFormatted != null ? timeFormatted : "No Time: ")
                        + name
                        + ": " + (content != null ? content : "NO get id:" + serverId)
                        + (media != null && media.equals("7") ?"スタンプ" : media != null && media.equals("1") ? "写真" : (media != null && media.equals("2") ? "動画" : ""));
                if (timeEpochStr == null) {
                    saveUnresolvedIds(serverId, talkId, unresolvedFilePath);
                }
                testWriter.write("再取得" + logEntry);
                testWriter.newLine();


            }


            try (BufferedWriter clearWriter = new BufferedWriter(new FileWriter(unresolvedFile))) {
                clearWriter.write("");
            }

        } catch (IOException e) {
            XposedBridge.log("IOException occurred while resolving and saving unresolved IDs: " + e.getMessage());
        }
    }


    private String formatMessageTime(String timeEpochStr) {
        if (timeEpochStr == null) return null;
        long timeEpoch = Long.parseLong(timeEpochStr);
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault());
        return sdf.format(new Date(timeEpoch));
    }

}'

areteruhiro commented:
かなり無理矢理ですが、対応させました。

areteruhiro commented:
重大なバグが発生しました。

修正するまで、マージしないでください。

areteruhiro commented:
再起動後に開くと、無限に
idが書き込まれるようです。
同じidが書かれないようにします。

areteruhiro commented:
対応しました。

s1204IT commented:
最終修正が完了次第、マージします。

areteruhiro commented:
すまません、前回した作業忘れてしまって、意味のわからないマージしました・

areteruhiro commented:
一回閉じてひらきなおしますコミット履歴
ヤバすぎるので

s1204IT commented:
マージの準備が出来たら、ファイル名の修正(`.java`を付ける)や、改行の数・ネストの空白の修正等を行ってからレビュー状態にしてください。

areteruhiro commented:
レビューリクエスト勝手に送信されちゃうんですよね、

.java
付けないといけないの
さきほど、初めてしったので、すみません、、気をつけます。

areteruhiro commented:
AndroidStudioだと、ファイルつくるときにファイルの拡張子決めるので。
